src/tools/compiletest/src/header.rs

```diff
@@ -925,8 +960,70 @@ fn iter_header(
             return;
 
         // First try to accept `ui_test` style comments
-        } else if let Some((header_revision, directive)) = line_directive(comment, ln) {
-            it(HeaderLine { line_number, original_line, header_revision, directive });
+        } else if let Some((header_revision, original_directive_line)) = line_directive(comment, ln)
+        {
+            // Let Makefiles and non-rs files just get handled in the regular way.
+            if testfile.extension().map(|e| e != "rs").unwrap_or(true) {
+                it(HeaderLine {
+                    line_number,
+                    original_line,
+                    header_revision,
+                    directive: original_directive_line,
+                });
+                continue;
+            }
+
+            let directive_ln = original_directive_line.trim();
+
+            if let Some((directive_name, _)) = directive_ln.split_once([':', ' ']) {
+                if KNOWN_DIRECTIVE_NAMES.contains(&directive_name) {
+                    it(HeaderLine {
+                        line_number,
+                        original_line,
+                        header_revision,
+                        directive: original_directive_line,
+                    });
+                    continue;
+                } else {
+                    *poisoned = true;
+                    eprintln!(
+                        "error: detected unknown compiletest test directive `{}` in {}:{}",
+                        directive_ln,
+                        testfile.display(),
+                        line_number,
+                    );
+
+                    if let Some(suggestion) =
+                        find_best_match_for_name(KNOWN_DIRECTIVE_NAMES, directive_name, None)
+                    {
+                        eprintln!("help: did you mean `{}` instead?", suggestion);
+                    }
+                    return;
+                }
+            } else if KNOWN_DIRECTIVE_NAMES.contains(&directive_ln) {
+                it(HeaderLine {
+                    line_number,
+                    original_line,
+                    header_revision,
+                    directive: original_directive_line,
+                });
+                continue;
+            } else {
+                *poisoned = true;
+                eprintln!(
+                    "error: detected unknown compiletest test directive `{}` in {}:{}",
+                    directive_ln,
+                    testfile.display(),
+                    line_number,
+                );
+
+                if let Some(suggestion) =
+                    find_best_match_for_name(KNOWN_DIRECTIVE_NAMES, directive_ln, None)
+                {
+                    eprintln!("help: did you mean `{}` instead?", suggestion);
+                }
+                return;
+            }
```

Comment:
Can we write some unit tests for this logic (or maybe for the whole feature)?

Reply:
~~or maybe for the whole compiletest~~ But yeah, I'll think about how to test this / make this more testable.

Reply:
Simplified this logic and added some (integration) tests for `iter_header` and directives checking.